Performance improvements for IdDict

Zentrik opened the pull request:
The main improvement is from `get!` which no longer calculates the hash twice when it has to setindex and so is twice as fast. This is my first time working with the Julia C code so there might be some mistakes there but it seems to work fine for me.
This is useful for a global value numbering pass I've written.

Zentrik committed:
Performance improvements for IdDict

Remove old comment

Switch to atomic pointers

Fixes errors in ASAN as previously we did atomic stores to non atomic pointers

Split empty! change into seperate pr

Remove duplication in C code

Add tests

15% speed up for setindex!

Remove try catch

Signifcant performance improvement for path where error used to be caught.

Fix bug

Actually fix bug

Restore performance by not boxing key twice

Fix comment

Add comment about IdDict being mirrored in julia.h

Fix comment

Fix another comment

Allow writing NULL into iddict

Reverts the change of behaviour just in case.

Remove clutter

KristofferC commented on base/iddict.jl:
```diff
@@ -132,8 +132,8 @@ function delete!(d::IdDict{K}, @nospecialize(key)) where K
 
-function empty!(d::IdDict)
-    resize!(d.ht, 32)
+function empty!(d::IdDict; preserve_size=false)
```
In my opinion, this should be in a separate PR while all the non API performance improvement can stay here.

JeffBezanson commented:
Thanks, any speedup numbers to whet the appetite?

I'm not really comfortable duplicating all the lookup code, and the try-catch around the conversion is not a good pattern to use. This needs a different design. Maybe we can copy how `get!` works for `Dict`.

Zentrik commented:
Here's some performance numbers,
```jl
function f!(d)
    for x in 1:10^4
        d[x] = 1
    end
end

function g!(d)
   for x in 1:10^4
       get!(d, x, 1)
   end
end
```
On master
```jl
julia> versioninfo()
Julia Version 1.11.0-DEV.356
Commit 4768d7ab8fe (2023-08-28 12:54 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 16 × AMD Ryzen 7 1700 Eight-Core Processor
  WORD_SIZE: 64
  LLVM: libLLVM-15.0.7 (ORCJIT, znver1)
  Threads: 1 on 16 virtual cores

julia> @btime f!(d) setup=(d=IdDict{Int, Int}(); sizehint!(d, 10^4))
  409.015 μs (9489 allocations: 148.27 KiB)

julia> @btime g!(d) setup=(d=IdDict{Int, Int}(); sizehint!(d, 10^4))
  639.763 μs (18978 allocations: 296.53 KiB)
```
On this pr
```jl
@btime f!(d) setup=(d=IdDict{Int, Int}(); sizehint!(d, 10^4))
  324.515 μs (9489 allocations: 148.27 KiB)

@btime g!(d) setup=(d=IdDict{Int, Int}(); sizehint!(d, 10^4))
  330.156 μs (9489 allocations: 148.27 KiB)
```

oscardssmith commented on base/iddict.jl:
```diff
@@ -83,18 +84,31 @@ function sizehint!(d::IdDict, newsz)
-    if !(val isa V) # avoid a dynamic call
-        val = convert(V, val)::V
-    end
+    keyindex = ccall(:jl_eqtable_keyindex, Cssize_t, (Any, Any), d, key)
```
why does this need to be implemented in C?

Zentrik replied:
I'm not sure how to call `jl_table_assign_bp` directly from Julia as it can change `d.ht`.

vtjnash replied:
There used to be ways, but not anymore with the generational GC. The usual way now to implement this is to switch the return object around as such:
```suggestion
    keyindex = Ref(UInt(0))
    d.ht = ccall(:jl_table_assign_bp, Any, (Any, Any, Ref{UInt}), d.ht, key, keyindex)
    keyindex = keyindex[]
```

vtjnash replied:
Can you change this around still also? It will avoid us from needing to reflect the iddict struct into C

Zentrik replied:
Just to note, I believe this would remove the ~20% speedup (see comments above) we get from not having to heap allocate for `setindex!`. Happy to make this change if that's fine.

oscardssmith commented on test/dict.jl:
```diff
@@ -701,6 +701,79 @@ end
     # test rethrow of error in ctor
     @test_throws DomainError   IdDict((sqrt(p[1]), sqrt(p[2])) for p in zip(-1:2, -1:2))
+
+    h = IdDict()
```
These tests look like they will be pretty slow. Can you crank down the numbers?

Zentrik replied:
Yep sorry, I just blindly copied from the tests above.

Zentrik committed:
Reduce loop iterations in tests

Zentrik committed:
Move code for calculating key index into Julia

Added inbounds as dict.jl seems to use them as well

Zentrik committed:
Decrease inlining cost of key indexing function

Unfortunately, cost is ~106 now so still too high, #51912 is tracking issue about high cost.

Lowers inlining cost of setproperty as we know rhs isa Memory{Any}

Zentrik committed:
Force inlining

About 15% faster on microbenchmark, `f!(d)` example in pr comments

vtjnash commented on base/iddict.jl:
```diff
@@ -83,18 +83,39 @@ function sizehint!(d::IdDict, newsz)
+    ref = RefValue{Ptr{Memory{Any}}}(pointer_from_objref(ht))
+    # # keyindex - where a key is stored, or -pos if the key was not present and was inserted at pos
+    keyindex = ccall(:jl_table_assign_bp, Cssize_t, (Ptr{Ptr{Memory{Any}}}, Any, Any, Cint), ref, key, C_NULL, 0)
+    d.ht = unsafe_pointer_to_objref(ref[])::Memory{Any}
```
This is a violation of the memory model as the value is not rooted (and cannot be rooted again after pointer_from_objref destroys that information chain)

Zentrik committed:
Revert "Force inlining"

This reverts commit 7eb2ed6a93c0722015d2572ab16b18e2d0ba50b4.

Zentrik committed:
Revert "Decrease inlining cost of key indexing function"

This reverts commit 519dfe75ad01c32c0bfbc90918edd2559cb0f9f9.

Zentrik committed:
Revert "Move code for calculating key index into Julia"

This reverts commit 7f2083310529daf49a7494b66cd37234b500081c.

Zentrik committed:
Add inbounds back in

Added inbounds as dict.jl seems to use them as well

vtjnash commented on base/iddict.jl:
```diff
@@ -176,16 +193,17 @@ function get(default::Callable, d::IdDict{K,V}, @nospecialize(key)) where {K, V}
+    keyindex, inserted = ht_keyindex!(d, key)
+
+    if inserted
         val = default()
```
It is not legal to use `_setindex` after this call

Zentrik replied:
You mean the call to `default` right? Presumably it's not legal because the dictionary might be mutated within the call?
I'll add a test for that then and revert the changes to this function.

vtjnash replied:
Yes, thanks

Zentrik committed:
Add trailing new line

Zentrik committed:
Have consistent spacing for type parameters

Zentrik committed:
Revert changes to get! with callable and add test.

Fixes bug if for example the call to `default` mutated the dict.

Zentrik committed:
Remove unused argument from `_setindex!` function

Zentrik commented:
CI looks good, ASAN just timed out I think.

vtjnash commented on base/iddict.jl:
```diff
@@ -83,18 +84,31 @@ function sizehint!(d::IdDict, newsz)
 end
 
-function setindex!(d::IdDict{K,V}, @nospecialize(val), @nospecialize(key)) where {K, V}
+function ht_keyindex!(d::IdDict{K,V}, @nospecialize(key)) where {K, V}
```
This does not appear to follow the API for `ht_keyindex` particularly well. It is closer to `ht_keyindex2!`, though with minor deviations in the return that it looks like could be changed here to match?

vtjnash commented on base/iddict.jl:
```diff
@@ -83,18 +84,31 @@ function sizehint!(d::IdDict, newsz)
+    if !(val isa V) # avoid a dynamic call
+        val = convert(V, val)::V
+    end
+    _setindex!(d, val, keyindex, inserted)
```
Following `dict.jl` API, the `@inbounds` should be here rather than inside `_setindex!` (which gets marked with `@propagate_inbounds`

vtjnash commented on base/iddict.jl:
```diff
@@ -83,18 +84,31 @@ function sizehint!(d::IdDict, newsz)
+function setindex!(d::IdDict{K,V}, @nospecialize(val), @nospecialize(key)) where {K, V}
+    keyindex, inserted = ht_keyindex!(d, key)
+    if !(val isa V) # avoid a dynamic call
+        val = convert(V, val)::V
```
There seems to be a discrepancy here in the state of the IdDict during this convert call or if this convert failed. The `ht_keyindex!` call has fully inserted the key, but not the value. For `Dict`, neither has occurred yet, and all insertion and accounting happens at once in `_setindex!`. But in IdDict here, it looks like half of the insertion occurs in `ht_keyindex!` (effectively inserting it in the deleted state) but then the final accounting update for it doesn't happen until after the `_setindex!` call later in this method. Do we update `ndel` over this call so the value is consistent?

There also may need to be an `age` counter added to the IdDict, so it can detect concurrent modifications that happen in this `convert` call (or move this convert call back to the top of the function)

Zentrik committed:
Conform to dict API and deal with `convert` failing

Conform to dict API, by moving `@inbounds` and changing `ht_keyindex!`.
Deal with failure case when `convert` calls fails.

TODO: think about concurrent modifications, should only be a problem in `get!`. Add tests for `convert` failing.

vtjnash commented on base/iddict.jl:
```diff
@@ -153,16 +171,24 @@ end
+
+    if keyindex < 0
+        # If convert call fails we need the key to be deleted
+        @inbounds d[-keyindex] = nothing
```
```suggestion
```

Zentrik replied:
Does the deleted state not require they key to be `nothing`? Though this prevents a segfault so I'm probably missing something.

Zentrik replied:
I think I get it now.

Zentrik replied:
Nevermind, the segfault was because I meant to write `d.ht[-keyindex] = nothing`. My concern is if the `convert` fails we have a valid key inserted and the value is NULL. This will prevent us from ever updating the value as we will hit this assert https://github.com/JuliaLang/julia/blob/e22db88127197cb3236e350cdb4ccbfff6cde232/src/iddict.c#L63-L69

I can move the key updating code out of `ht_keyindex2!` if that helps.

Zentrik committed:
Fix bug in writing to dict

Zentrik committed:
Don't delete key

Caused segfaults when ht_keyindex2! caused a rehash

Zentrik committed:
Add age field to fix get!

The age field detects concurrent modifications in the convert or default calls. Copied from Dict.

Zentrik committed:
Do not pass IdDict into C function

Whilst this seemed to be a performance improvement when this pr was created, it no longer is.

Zentrik committed:
Move around some code

Zentrik committed:
Add an inline

Zentrik committed:
Simplify by moving key write to Julia

Zentrik commented on base/iddict.jl:
```diff
@@ -165,17 +165,47 @@ function get(default::Callable, d::IdDict{K,V}, @nospecialize(key)) where {K, V}
-        return val
-    else
+        if d.age != age0
+            @inline setindex!(d, val, key)
```
```suggestion
            setindex!(d, val, key)
```

Zentrik committed:
Remove superfluous inline

Zentrik commented on base/iddict.jl:
```diff
@@ -165,17 +165,47 @@ function get(default::Callable, d::IdDict{K,V}, @nospecialize(key)) where {K, V}
+    return keyindex[]
+end
+
+@propagate_inbounds function _setindex!(d::IdDict{K,V}, val::V, key::k, keyindex::Int) where {K, V}
```
```suggestion
@propagate_inbounds function _setindex!(d::IdDict{K,V}, val::V, key:K, keyindex::Int) where {K, V}
```

Zentrik committed:
Fix typo

Zentrik committed:
Fix typo

adienes commented:
still want this?
